From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: matthewperiut <matthewperiut@gmail.com>
Date: Wed, 23 Sep 2026 21:34:18 -0400
Subject: [PATCH 18/19] kk: allow rendering to linear color images (not as
 input attachments)

Metal renders to buffer-backed linear textures; only input attachments
(implemented as arrays) can't be linear. Needed for dma-buf sharing in VMs
(virglrenderer on macOS shares linear images only).
---
 src/kosmickrisp/vulkan/kk_image.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/kosmickrisp/vulkan/kk_image.c b/src/kosmickrisp/vulkan/kk_image.c
index 5b1a685..51ba560 100644
--- a/src/kosmickrisp/vulkan/kk_image.c
+++ b/src/kosmickrisp/vulkan/kk_image.c
@@ -61,9 +61,11 @@ kk_get_image_plane_format_features(struct kk_physical_device *pdev,
       features |= VK_FORMAT_FEATURE_2_SAMPLED_IMAGE_DEPTH_COMPARISON_BIT;
    }
 
-   /* We disable A8 format due to lower blend pass issues */
-   if (va_format->color && tiling != VK_IMAGE_TILING_LINEAR &&
-       vk_format != VK_FORMAT_A8_UNORM) {
+   /* We disable A8 format due to lower blend pass issues. Linear (buffer
+    * backed) textures can be render targets in Metal too: dma-buf sharing in
+    * virtual machines (virglrenderer on macOS) needs renderable linear images.
+    */
+   if (va_format->color && vk_format != VK_FORMAT_A8_UNORM) {
       features |= VK_FORMAT_FEATURE_2_COLOR_ATTACHMENT_BIT;
       features |= VK_FORMAT_FEATURE_2_BLIT_DST_BIT;
       // TODO_KOSMICKRISP Support snorm formats once the following spec issue is
@@ -280,7 +282,7 @@ kk_GetPhysicalDeviceImageFormatProperties2(
     */
    if (pImageFormatInfo->tiling == VK_IMAGE_TILING_LINEAR &&
        (pImageFormatInfo->usage &
-        (VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT |
+        (VK_IMAGE_USAGE_INPUT_ATTACHMENT_BIT |
          VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT)))
       return VK_ERROR_FORMAT_NOT_SUPPORTED;
 
-- 
2.54.0 (Apple Git-157)

